lib/web/autoupdate_common.go

```diff
@@ -0,0 +1,220 @@
+/*
+ * Teleport
+ * Copyright (C) 2024  Gravitational, Inc.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package web
+
+import (
+	"context"
+	"strings"
+
+	"github.com/gravitational/trace"
+
+	autoupdatepb "github.com/gravitational/teleport/api/gen/proto/go/teleport/autoupdate/v1"
+	"github.com/gravitational/teleport/api/types/autoupdate"
+	"github.com/gravitational/teleport/lib/automaticupgrades"
+)
+
+// autoUpdateAgentVersion returns the version the agent should install/update to based on
+// its group and updater UUID.
+// If the cluster contains an autoupdate_agent_rollout resource from RFD184 it should take precedence.
+// If the resource is not there, we fall back to RFD109-style updates with channels
+// and maintenance window derived from the cluster_maintenance_config resource.
+// Version returned follows semver without the leading "v".
+func (h *Handler) autoUpdateAgentVersion(ctx context.Context, group, updaterUUID string) (string, error) {
+	rollout, err := h.cfg.AccessPoint.GetAutoUpdateAgentRollout(ctx)
+	switch {
+	case err == nil:
+		return getVersionFromRollout(rollout, group, updaterUUID)
+	case trace.IsNotFound(err):
+		return getVersionFromChannel(ctx, h.cfg.AutomaticUpgradesChannels, group)
+	default:
+		return "", trace.Wrap(err, "Failed to get auto-update rollout")
```

Comment:
nit: not familiar with trace.Wrap yet, but is it expected to have a capitalized "Failed to"? In "regular" %w wrapping, it would just be "get auto-update rollout".